mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala

```diff
@@ -479,9 +492,16 @@ class Word2VecModel private[mllib] (
    */
   def findSynonyms(vector: Vector, num: Int): Array[(String, Double)] = {
     require(num > 0, "Number of similar words should > 0")
-    // TODO: optimize top-k
-    val fVector = vector.toArray.map(_.toFloat)
-    model.mapValues(vec => cosineSimilarity(fVector, vec))
+
+    val numWords = wordVectors.numRows
+    val cosineVec = Vectors.zeros(numWords).asInstanceOf[DenseVector]
+    BLAS.gemv(1.0, wordVectors, vector.asInstanceOf[DenseVector], 0.0, cosineVec)
+
+    // Need not divide with the norm of the given vector since it is constant.
+    val updatedCosines = indexedModel.map { case (_, ind) =>
```

Comment:
I'm not sure if an iterator over a Map has guarantees about the ordering of the elements.  It's probably better not to assume that will remain stable.